feat: Add helper function to verify bounds of timestamp search

OussamaSaoudi opened the pull request:
## What changes are proposed in this pull request?

 This PR introduces:
  - `timestamp_to_version`: Core function that converts a timestamp to a Delta version using binary search,
  supporting both GreatestLower (latest version at or before timestamp) and LeastUpper (first version at
  or after timestamp) bounds
  - get_timestamp_search_bounds: Determines the appropriate search strategy based on In-Commit Timestamp (ICT) enablement, handling the transition between file modification timestamps and ICT
  - TimestampSearchBounds: Enum representing the four possible search scenarios (exact match, ICT range,
  file modification range, or mixed)
  - Simplified LogHistoryError types with consolidated error variants

  The implementation correctly handles tables that transition from file modification timestamps to
  In-Commit Timestamps mid-history, ensuring searches are constrained to the appropriate timestamp domain.


**Stack:**
- PR #899 (this PR)
- PR #900

## How was this change tested?
Unit tests cover:
- Binary search with exact matches and no matches
- Duplicate value handling (first/last occurrence)
- Edge cases (empty arrays, out of range values)
- Error propagation from key functions

OussamaSaoudi commented on kernel/src/history_manager/mod.rs:
```diff
@@ -134,6 +147,152 @@ fn read_in_commit_timestamp(
+    timestamp: Timestamp,
+    bound: Bound,
+) -> Result<Version, LogHistoryError> {
+    require!(timestamp >= 0, LogHistoryError::InvalidTimestamp(timestamp));
```
Oh I didn't even make the connection that negative values can represent time before 1970! I'll fix that.

OussamaSaoudi commented:
aside: I realized that the TimestampOutOfRangeError is now expensive to build without our cache of values. I can take that up as a followup

OussamaSaoudi commented on kernel/src/history_manager/error.rs:
```diff
@@ -41,3 +45,66 @@ pub enum LogHistoryError {
+    pub(crate) fn try_new<'a>(
+        commit_range: &'a [ParsedLogPath],
+        timestamp: Timestamp,
+        key_fun: impl Fn(&'a ParsedLogPath) -> Result<i64, LogHistoryError>,
```
good catch!

scovich commented on kernel/src/history_manager/error.rs:
```diff
@@ -42,3 +46,66 @@ pub enum LogHistoryError {
+
+impl TimestampOutOfRangeError {
+    pub(crate) fn try_new<'a>(
+        commit_range: &'a [ParsedLogPath],
+        timestamp: Timestamp,
+        key_fun: impl Fn(&'a ParsedLogPath) -> Result<Timestamp, LogHistoryError>,
```
Remind me why these args need lifetimes? We don't keep any references that outlive the function call?
(I feel like I may have asked this before and received an answer, but I can't find it anywhere)

OussamaSaoudi replied:
removed them 👍

scovich commented on kernel/src/history_manager/error.rs:
```diff
@@ -42,3 +46,66 @@ pub enum LogHistoryError {
+impl Display for TimestampOutOfRangeError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.bound {
+            Bound::LeastUpper => {
+                write!(
```
aside: I'm surprised fmt added curly braces here? I would have expected e.g.
```rust
Bound::LeastUpper => write!(
    ...
),
```

scovich commented on kernel/src/history_manager/mod.rs:
```diff
@@ -124,6 +136,146 @@ fn read_in_commit_timestamp(
+        log_segment
+            .ascending_commit_files
+            .binary_search_by(|x| x.version.cmp(&ict_version))
+            .unwrap_or_else(|idx| idx)
```
Maybe worth expanding the above comment to explain how we use the Err result here (why it's correct)?

scovich replied:
Also, why pull it out as a closure like this? Wouldn't it be just as compact if we fold it into its only callsite in the match below?

OussamaSaoudi replied:
removed closure.

scovich reviewed (approved):
LGTM, few nits/questions to consider

OussamaSaoudi committed:
feat: add low-level timestamp-to-version conversion

Add core infrastructure for timestamp-based time travel:
- timestamp_to_version: converts timestamp to version using binary search
- get_timestamp_search_bounds: determines search strategy based on ICT enablement
- TimestampSearchBounds: enum for search scenarios
- Simplified LogHistoryError types (collapsed TimestampOutOfRangeError into enum)
- Made Bound public for use in error types

Includes debug tracing for search bounds and ICT reads.

Co-authored-by: Isaac

OussamaSaoudi committed:
feat: add debug assertions for search bounds validation

Add debug_assert! statements to verify lo and hi indices are within valid
bounds before array access in timestamp_to_version.

Co-authored-by: Isaac

OussamaSaoudi committed:
feat: add monotonization for file modification timestamp searches

File modification timestamps are not guaranteed to be monotonically
increasing due to clock skew or filesystem limitations. This change:

- Adds linear_search_file_mod_timestamps() which monotonizes timestamps
  during search (if ts <= prev_ts, adjust to prev_ts + 1)
- Uses binary search only for ICT-based searches (ICT is guaranteed
  monotonic by the Delta protocol)
- Uses linear scan with monotonization for file modification searches

This matches the behavior of delta-kernel-java's DeltaHistoryManager.

Co-authored-by: Isaac

OussamaSaoudi committed:
docs: add monotonization trade-offs comment

Document the trade-offs of file modification timestamp monotonization
in linear_search_file_mod_timestamps, including edge cases near
retention boundaries and the relationship with ICT as the long-term
solution.

Co-authored-by: Isaac

OussamaSaoudi committed:
chore: revert unrelated formatting change

Co-authored-by: Isaac

OussamaSaoudi committed:
refactor: make Bound internal by storing reason string in error

Change TimestampOutOfRange error to store a reason string instead of
the Bound enum, keeping Bound as an internal implementation detail.

Co-authored-by: Isaac

OussamaSaoudi committed:
refactor: make LogHistoryError::Internal source optional

Allow creating internal errors with just a context message (no underlying
source error). This simplifies error creation for cases where there's no
wrapped error.

Co-authored-by: Isaac

OussamaSaoudi committed:
fix: make history_manager module pub(crate) with allow(unused)

The low-level timestamp APIs are internal implementation details.
- Change history_manager module from internal-api conditional to pub(crate)
- Add #![allow(unused)] to all module files since they're unused until
  the high-level PR exposes them publicly

Co-authored-by: Isaac

OussamaSaoudi committed:
docs: add protocol-based documentation for ICT search bounds match arms

Document each match arm in get_timestamp_search_bounds with protocol
reasoning for future reviewers (human or otherwise).

Co-authored-by: Isaac

OussamaSaoudi committed:
refactor: change non-monotonic timestamp log from warn to trace

Non-monotonic file modification timestamps are expected in some
environments and don't indicate a problem - they're handled correctly
by monotonization. Downgrade to trace level to avoid noisy logs.

Co-authored-by: Isaac
